src/msg/msg_types.h

```diff
@@ -217,8 +217,8 @@ struct entity_addr_t {
     }
   };
 
-  __u32 type;
-  __u32 nonce;
+  ceph_le32 type;
+  ceph_le32 nonce;
   union {
```

Comment:
How about the `sockaddr`-based members of the union?

Reply:
Please correct me if I wrong:
The IP addresses are in network byte order in both LE or BE machines. Then the data are serialized into the bufferlist which is BE, so the type and nonce are hton-ed while the sockaddr data are copied in BE format.

Reply:
If these values are always in the network byte order, then obviously everything should be fine. What worries me is whether this assumptions truly stands as a ctor of `entity_addr_t` does `ntohs()`:

```cpp
  explicit entity_addr_t(const ceph_entity_addr &o) {
    type = o.type;
    nonce = o.nonce;
    memcpy(&u, &o.in_addr, sizeof(u));
#if !defined(__FreeBSD__)
    u.sa.sa_family = ntohs(u.sa.sa_family);
#endif
  }
```

Reply:
If I understand correctly the `*_family` is used to select the `AF_INET` or `AF_INET6` structure. It is encoded from host to net (encode()) and decoded from net to the host (decode()). The data in the `ceph_entity_addr::sockaddr_storage` is in network storage so the `*_family` has to be converted.

Reply:
> If I understand correctly the *_family is used to select the AF_INET or AF_INET6 structure. It is encoded from host to net (encode()) and decoded from net to the host (decode()).

I believe what happens in `encode()` is irrelevant. What matters is the behavior of the comparison operator which, in my understanding, operates on data being at least partially (even after your patch) in the host byte order. Please take a look on: https://gist.github.com/rzarzynski/4493f9bb6e28518006fdee0966cc10dd#the-comparison-operator.

Reply:
Letmeread

Reply:
Thank you Radoslaw very much for the doc! Do you think the sorting can be done using an int as in the `entity_name_t`? Btw, some other structs have ordering by memcmp which potentially should give difference for the LE and BE platforms...

Reply:
> Do you think the sorting can be done using an int as in the entity_name_t

Not sure whether I got the idea. Could you please rephrase / exemplify? What I think for now is that we need to preserve the current sorting order where it currently works due to upgrades.

> Btw, some other structs have ordering by memcmp which potentially should give difference for the LE and BE platforms...

Ouch, I feel we've opened a Pandora's can. I'm curious how many bugs was inside ;-).

Reply:
I mean instead of comparing memory to compare the `ceph_le32 nonce`. Many other structs use integers to make comparison. Should be safe at the BE and LE.

Luckilly not many:
`inline bool operator<(const ACLID& l, const ACLID& r)`.

Reply:
Just `nonce` or everything about to `int` / to-address conversions?

Reply:
Need to grep the whole code if such a sorting is appropriate. It would be great...

Reply:
What is dangerous is that the sorting (depended on the address) is used in probing/polling/etc. and performance can be degraded.
It looks like there is no problem with this structure especially in the containers.
May be you have another information on those structs and containers.